work

odscjames opened the pull request:
-   Work out number of errors in Scrapy log file
-   Consider number of errors in decision to backup or not
-   Don't backup subsets (sample, date filters)
-   Don't backup if data files missing
-   Add Sentry
-  Add test run option
-  Delete files we make and log and data files

jarofgreen committed:
work: errors, subsets, data missing, sentry

Work out number of errors in Scrapy log file
Consider number of errors in decision to backup or not
Don't backup subsets (sample, date filters)
Don't backup if data files missing
Add Sentry
Don't check deleted collections

odscjames commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Move Archive to S3 and decommission server](https://trello.com/c/LSzAt3nw/234-move-archive-to-s3-and-decommission-server)

odscjames commented:
Making draft while adding more things; will take out of draft when ready for review

jarofgreen committed:
cli: Add test run option

jarofgreen committed:
work: Delete local files we create

jarofgreen committed:
work: Delete data files and log files

jarofgreen committed:
work: Delete files downloaded from S3 after use

odscjames commented:
@jpmckinney Now ready for review - thanks

odscjames commented:
@jpmckinney Can you look at this? Thanks

jpmckinney commented on manage.py:
```diff
@@ -25,17 +26,22 @@ def cli():
 
 @cli.command()
-def archive():
+@click.option('-t', '--test-run', is_flag=True)
```
Change to `-n` and `--dry-run`, which is what many *nix commands use.

odscjames replied:
Done

jpmckinney commented on ocdskingfisherarchive/archive.py:
```diff
@@ -11,35 +12,49 @@ def __init__(self, config, database_archive, database_process, s3):
+            self.logger.info(
+                "Collection " + str(collection.database_id) + " result: " + ("Archive" if should_archive else "Leave")
+            )
+            print(
+                "Collection " + str(collection.database_id) + " result: " + ("Archive" if should_archive else "Leave")
+            )
```
Instead of both logging and printing, you should just add appropriate handlers.

odscjames replied:
If we are sure we want logging messages to go to console, then yes. However I remembered some chat about minimising noisy cron's, which would mean we don't. In that case having a print is good, so the operator can see results straight away without having to then go and look in the log files. That was my thinking behind putting the print in.

jpmckinney commented on ocdskingfisherarchive/archive.py:
```diff
@@ -11,35 +12,49 @@ def __init__(self, config, database_archive, database_process, s3):
+            )
         else:
-            self.database_archive.set_state_of_collection_id(collection.database_id, 'DO NOT ARCHIVE')
+            if should_archive:
```
Switch to `elif` to avoid unnecessary hierarchy.

odscjames replied:
Done

jpmckinney commented on ocdskingfisherarchive/archive.py:
```diff
@@ -11,35 +12,49 @@ def __init__(self, config, database_archive, database_process, s3):
 
-        # TODO check data files exist
+        if not collection.get_data_files_exist():
+            self.logger.debug('Not Archiving Cos data files do not exist')
```
Please use correct English "because" not "Cos" throughout.

odscjames replied:
Done

jpmckinney commented on ocdskingfisherarchive/archive.py:
```diff
@@ -51,7 +66,12 @@ def should_we_archive_collection(self, collection):
+            # (But we may not have an errors count for one of the things we are comparing)
+            if collection.has_errors_count() and exact_archived_collection.has_errors_count() and \
+                    collection.get_errors_count() > exact_archived_collection.get_errors_count():
+                self.logger.debug('Not Archiving Cos An archive exists with less errors')
```
Use "fewer errors" not "less errors" throughout.

odscjames replied:
Done

jpmckinney commented on ocdskingfisherarchive/archive.py:
```diff
@@ -11,35 +12,49 @@ def __init__(self, config, database_archive, database_process, s3):
+        should_archive = self.should_we_archive_collection(collection)
+        if test_run:
+            self.logger.info(
+                "Collection " + str(collection.database_id) + " result: " + ("Archive" if should_archive else "Leave")
```
Prefer "Skip" to "Leave"

odscjames replied:
Done

jpmckinney commented on ocdskingfisherarchive/archive.py:
```diff
@@ -84,8 +104,14 @@ def should_we_archive_collection(self, collection):
+            # Clean: If the local directory has fewer errors, and greater or equal bytes, replace the remote directory.
+            # (But we may not have an errors count for one of the things we are comparing)
+            if collection.has_errors_count() and last_archived_collection.has_errors_count() and \
+                    collection.get_errors_count() < last_archived_collection.get_errors_count() and \
```
Should be `<=`, e.g. both have 0 errors but one is bigger.

odscjames replied:
Done with new test

jpmckinney commented on ocdskingfisherarchive/archive.py:
```diff
@@ -84,8 +104,14 @@ def should_we_archive_collection(self, collection):
+                    collection.get_errors_count() < last_archived_collection.get_errors_count() and \
+                    collection.get_size_of_data_folder() >= last_archived_collection.get_data_size():
+                self.logger.debug('Archiving Cos An archive exists with older year/month and ' +
+                                  'local collection has less errors and same or bigger size')
```
Use "greater or equal size" not "same or bigger size".

odscjames replied:
Done

jpmckinney commented on ocdskingfisherarchive/archive.py:
```diff
@@ -84,8 +104,14 @@ def should_we_archive_collection(self, collection):
+            if collection.has_errors_count() and last_archived_collection.has_errors_count() and \
+                    collection.get_errors_count() < last_archived_collection.get_errors_count() and \
+                    collection.get_size_of_data_folder() >= last_archived_collection.get_data_size():
+                self.logger.debug('Archiving Cos An archive exists with older year/month and ' +
```
Use correct capitalization (many words after "Cos" throughout are incorrectly capitalized).

odscjames replied:
Done

jpmckinney commented on ocdskingfisherarchive/archive.py:
```diff
@@ -137,7 +163,15 @@ def archive_collection(self, collection):
         self.s3.remove_staging_file(s3_directory + '/data.tar.lz4')
 
-        # TODO Delete data files
+        # Delete local files we made
```
I assume the lines before these will raise an exception if they fail? We don't want to reach this code if anything above failed.

odscjames replied:
Yes, there are very few try: blocks and only 3 relevant ones that don't just log and re-raise. Those 3 are very specific and limited (1 is a 404 when getting from S3 and the other 2 are problems parsing dicts in a log file). A discussion for later about opening that up, but to start with I'd just let it crash and we'll examine cases as they arise.

jpmckinney commented on ocdskingfisherarchive/archived_collection.py:
```diff
@@ -35,8 +36,9 @@ def _load(s3, source_id, year, month):
         filename = s3.get_file(remote_filename)
         if filename:
             with open(filename) as fp:
-                return ArchivedCollection(json.load(fp), year, month)
-            # TODO delete the file to save our /tmp space
+                archived_collection = ArchivedCollection(json.load(fp), year, month)
+            os.unlink(filename)
```
It seems a bit silly to write to a temporary file and then unlink it. If you just need to read it into memory, add a method to the s3 module that does:

    stream = io.StringIO()
    object.download_fileobj(stream)

Then you can do `json.load(stream)`

odscjames replied:
I looked into this; the above code doesn't work and it started to get complex; I tried switching to get_object which just gives you a streaming object back but oddly it appears to be difficult to catch 404's well due to boto's use of error factories. Let's issue this for later.

jpmckinney commented on ocdskingfisherarchive/collection.py:
```diff
@@ -13,14 +13,21 @@ def __init__(self, config, database_id, source_id, data_version):
             'database_id': self.database_id,
             'data_md5': self.get_md5_of_data_folder(),
             'data_size': self.get_size_of_data_folder(),
+            'scrapy_log_file_found': (True if self._scrapy_log_file else False),
```
Never do `True if condition else False` – it's unnecessarily complicated:

```suggestion
            'scrapy_log_file_found': bool(self._scrapy_log_file),
```

odscjames replied:
Done

jpmckinney commented on ocdskingfisherarchive/collection.py:
```diff
@@ -60,16 +67,46 @@ def get_size_of_data_folder(self):
+                   self.data_version.strftime("%Y%m%d_%H%M%S")
+
+        # It must exist and be a directory
+        if not os.path.exists(dir_name) or not os.path.isdir(dir_name):
```
`isdir` is enough (things that don't exist aren't directories). This occurs multiple times.

```suggestion
        if not os.path.isdir(dir_name):
```

jpmckinney replied:
Actually, you can just `return os.path.isdir(dir_name)`. Never do:

```python
if not condition:
    return False

return True
```

It's overly indirect.

odscjames replied:
I structured it if/return/return on the guess we might have some more checks later to go in the middle (ie does the directory actually have some .json files in it, or is it empty?) but easy to change later - all done.

jpmckinney commented on ocdskingfisherarchive/collection.py:
```diff
@@ -60,16 +67,46 @@ def get_size_of_data_folder(self):
+        return self._data_size
+
+    def get_data_files_exist(self):
+        dir_name = self.config.directory_data + '/' + self.source_id + '/' + \
```
Always use `os.path.join` for filesystem paths, to avoid cross-platform issues. Do this throughout.

(S3 paths aren't filesystem paths, so continue to use `/`, though `'/'.join(['foo', 'bar'])` is less error-prone (and incidentally way faster) than `+`.

odscjames replied:
We calculate data dir a lot, I actually moved it to a separate function.

odscjames replied:
All done

jpmckinney commented on ocdskingfisherarchive/collection.py:
```diff
@@ -60,16 +67,46 @@ def get_size_of_data_folder(self):
+
+    def _cache_scrapyd_log_file_info(self):
+        if self._scrapy_log_file_name is not None:
+            return self._scrapy_log_file_name
```
Why is there a return value here but not later?

odscjames replied:
Whoops, historical reasons. Done.

jpmckinney commented on ocdskingfisherarchive/collection.py:
```diff
@@ -60,16 +67,46 @@ def get_size_of_data_folder(self):
+        return True if self._scrapy_log_file else False
+
+    def get_errors_count(self):
+        self._cache_scrapyd_log_file_info()
```
This is very repetitive. Instead, you can do:

```python
@property
def scrapy_log_file(self):
    self._cache_scrapyd_log_file_info()
    return self._scrapy_log_file
```

That way, these methods can just call `self.scrapy_log_file.get_...` instead of calling the cache method first.

You can do the same for `scrapy_log_file_name`.

odscjames replied:
Done

jpmckinney commented on ocdskingfisherarchive/collection.py:
```diff
@@ -94,7 +131,32 @@ def write_data_file(self):
+            data_dir = self.config.directory_data + '/' + self.source_id + '/' + \
+                self.data_version.strftime("%Y%m%d_%H%M%S")
+            # We use os.system here so we know the exact command so we can set up sudo correctly
+            return1 = os.system('sudo -u ocdskfs /bin/rm -rf '+data_dir)
```
Why not just set things up so that the ocdskfs user runs the archive command? Seems like a lot of trouble otherwise (changes to the sudo command require changes to deploy...).

odscjames replied:
So this is actually matching the current setup as closely as possible, to minimise changes on server. We already have sudo set up: https://github.com/open-contracting/deploy/blob/master/salt/ocdskingfisherarchive/archive.sudoers

And also noting  that we have to have some salt changes when deploying this (config file, logging config file, new pillar variables, pip); I've already done that initial work and can present it soon. But minimising changes for now seemed sensible.

jpmckinney commented on ocdskingfisherarchive/scrapy_log_file.py:
```diff
@@ -6,13 +7,79 @@
+
+    # Manual Processing
+
+    def _process_manually(self):
```
Why is this called `_process_manually`? Seems to be automated!

Since it's a long method, a docstring to say what it's doing would be good (counting the number or errors and extracting the spider arguments, from a quick read).

odscjames replied:
I called it manual as opposed to by the library, but I can see that is confusing. Done.

jpmckinney commented on ocdskingfisherarchive/scrapy_log_file.py:
```diff
@@ -6,13 +7,79 @@
+            self._process_manually()
+
+        # Older spider log files may not have this data, so make sure it can deal with that case.
+        return bool(self._spider_arguments.get('sample')) or \
```
More filters will be added in future, so please open an issue on Kingfisher Collect to document all current filters (to make this easier to maintain).

odscjames replied:
https://github.com/open-contracting/kingfisher-collect/issues/506

jpmckinney reviewed (commented):
Looks good. Would be good to do the small tidying in the comments, and create issues for anything longer.

jarofgreen committed:
cli: change test-run to dry-run

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
archive: Remove unnecessary hierarchy

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
archive: Spelling and Grammar in log messages

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
archive: Grammar in log messages

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
archive: Skip not Leave in messages

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
scrapy_log_file: Rename method and docstring

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
collection: _cache_scrapyd_log_file_info should not return anything

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
collection: Simpler isdir checks

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
collection: Simpler bool check

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
collection: Move common code to _get_data_dir_name, use os.path

jarofgreen committed:
collection: calculate S3 path with join

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
archive: Also backup if earlier collection has same errors and local is bigger

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
archive: Correct debug message (both grammer and correctness)

https://github.com/open-contracting/kingfisher-archive/pull/25

" fewer or equal errors" should have been in
8228dbf6a881e4c8bf179a8287692ab8c10af5ef

jarofgreen committed:
collection: use python @property for log file

https://github.com/open-contracting/kingfisher-archive/pull/25

jarofgreen committed:
s3: bug fix; create logger

jarofgreen committed:
database_process: Get collections in order

jpmckinney committed:
Use the same short flag for --dry-run as most other CLI tools

jpmckinney committed:
Use methods instead of repeating their bodies

jpmckinney reviewed (approved):
I added a couple commits and created issues for outstanding comments.

jpmckinney commented:
I'm going to merge so that I can do the other admin setting changes. Let me know if you have any issue with my two commits.
